vhdx/src/vhdx_io.rs

```diff
@@ -0,0 +1,235 @@
+            UnsupportedMode => write!(f, "Differencing mode is not supported yet"),
+        }
+    }
+}
+
+macro_rules! round_up {
+    ($n:expr, $align:expr) => {{
+        if $align > $n {
+            $align
+        } else {
+            let rem = $n % $align;
+            (($n / $align) + rem) * $align
+        }
+    }};
+}
+
+#[derive(Default)]
+struct Sector {
+    bat_index: u64,
+    free_sectors: u64,
+    free_bytes: u64,
+    file_offset: u64,
+    block_offset: u64,
+}
+
+impl Sector {
+    /// Translate secotor index and count of data in file to actual offsets and BAT index
+    pub fn new(
+        disk_spec: &DiskSpec,
+        bat: &[BatEntry],
+        sector_index: &u64,
+        sector_count: &u64,
+    ) -> Result<Sector> {
+        let mut sector = Sector::default();
+
+        sector.bat_index = *sector_index / disk_spec.sectors_per_block as u64;
+        sector.block_offset = *sector_index % disk_spec.sectors_per_block as u64;
+        sector.free_sectors = disk_spec.sectors_per_block as u64 - sector.block_offset;
+        if sector.free_sectors > *sector_count {
+            sector.free_sectors = *sector_count;
+        }
+
+        sector.free_bytes = sector.free_sectors * disk_spec.logical_sector_size as u64;
+        sector.block_offset *= disk_spec.logical_sector_size as u64;
+
+        let bat_entry = match bat.get(sector.bat_index as usize) {
+            Some(entry) => (*entry).0,
+            None => {
+                return Err(Error::InvalidBatIndex);
+            }
+        };
+        sector.file_offset = bat_entry & vhdx_bat::BAT_FILE_OFF_MASK;
+        if sector.file_offset != 0 {
+            sector.file_offset += sector.block_offset;
+        }
+
+        Ok(sector)
+    }
+}
+
+/// VHDx IO read routine: requires relative sector index and count for the desired data
+pub fn read(
+    f: &mut File,
+    buf: &mut [u8],
+    disk_spec: &DiskSpec,
+    bat: &[BatEntry],
+    mut sector_index: u64,
+    mut sector_count: u64,
+) -> Result<usize> {
+    let mut read_count: usize = 0;
+
+    while sector_count > 0 {
+        if disk_spec.has_parent {
+            return Err(Error::UnsupportedMode);
+        } else {
+            let sector = Sector::new(disk_spec, bat, &sector_index, &sector_count)?;
+
+            let bat_entry = match bat.get(sector.bat_index as usize) {
+                Some(entry) => (*entry).0,
+                None => {
+                    return Err(Error::InvalidBatIndex);
+                }
+            };
+
+            match bat_entry & vhdx_bat::BAT_STATE_BIT_MASK {
+                vhdx_bat::PAYLOAD_BLOCK_NOT_PRESENT
+                | vhdx_bat::PAYLOAD_BLOCK_UNDEFINED
+                | vhdx_bat::PAYLOAD_BLOCK_UNMAPPED
+                | vhdx_bat::PAYLOAD_BLOCK_UNMAPPED_V095
+                | vhdx_bat::PAYLOAD_BLOCK_ZERO => {}
+                vhdx_bat::PAYLOAD_BLOCK_FULLY_PRESENT => {
+                    // should we put lock here or in the calling function?
```

Comment:
You can remove this comment, the fact that this function expect a mutable reference onto the file means that the caller will have to make sure there's no concurrency.